doc/manual/source/store/drv.md

```diff
@@ -0,0 +1,311 @@
+Other system (like Git or IPFS) also store and transfer immutable data, but they don't concern themselves with *how* that data was created.
+
+This is where Nix distinguishes itself.
+*Derivations* represent individual build steps, and *deriving paths* are needed to refer to the *outputs* of those build steps before they are built.
+<!-- The two concepts need to be introduced together because, as described below, each depends on the other. -->
+
+## Store Derivation {#store-derivation}
+
+A derivation is a specification for running an executable on precisely defined input files to repeatably produce output files at uniquely determined file system paths.
+
+What is the natural Unix analog for a build step *in action*?
+Answer: a process that will eventually exit, leaving behind some output files.
+What is the natural way to *plan* such a step?
+An `execve` system call.
+
+A derivation consists of:
+
+ - A name
+
+ - A set of [*inputs*][inputs], a set of [deriving paths][deriving path]
+
+ - A map of [*outputs*][outputs], from names to other data
+
+ - The [process creation fields]: to spawn the arbitrary process which will perform the build step:
+
+
+   - The [*builder*][builder], a path to an executable
+
+   - A list of [arguments][args]
+
+   - A map of [environment variables][env].
+
+ - A two-component ["system" type][system] (e.g. `x86_64-linux`) where the executable is to run.
+
+[store derivation]: #store-derivation
+[inputs]: #inputs
+[input]: #inputs
+[outputs]: #outputs
+[output]: #outputs
+[process creation fields]: #process-creation-fields
+[builder]: #builder
+[args]: #args
+[env]: #env
+[system]: #system
+
+### Referencing derivations {#derivation-path}
+
+Derivations are always referred to by the [store path] of the store object they are encoded to.
+See the [encoding section](#derivation-encoding) for more details on how this encoding works, and thus what exactly what store path we would end up with for a given derivation.
+
+The store path of the store object which encodes a derivation is often called a *derivation path* for brevity.
+
+## Deriving path {#deriving-path}
+
+Deriving paths are a way to refer to [store objects][store object] that might not yet be [realised][realise].
+This is necessary because, in general and particularly for [content-addressed derivations][content-addressed derivation], the [store path] of an [output] is not known in advance.
+There are two forms:
+
+- [*constant*]{#deriving-path-constant}: just a [store path].
+  It can be made [valid][validity] by copying it into the store: from the evaluator, command line interface or another st    ore.
+
+- [*output*]{#deriving-path-output}: a pair of a [store path] to a [store derivation] and an [output] name.
+
+In pseudo code:
+
+```idris
+type OutputName = String
+
+data DerivingPath
+  = Constant { path : StorePath }
+  | Output {
+      drvPath : StorePath,
+      output  : OutputName,
+    }
+```
+
```

Comment:
```suggestion

Deriving paths are necessary because, in general and particularly for [content-addressed derivations][content-addressed derivation], the [store path] of an [output] is not known in advance.
We can use an output deriving path to refer to such an out, instead of the store path which we do not yet know.

```